src/etl/ETLSource.h

```diff
@@ -134,6 +142,34 @@ class ETLSource
         boost::asio::yield_context& yield) const = 0;
 };
 
+/// Hooks are executed with respect to the mutex provided
+struct ETLSourceHooks
+{
+    enum class Action { stop, proceed };
+
+    std::reference_wrapper<std::mutex> mtx_;
```

Comment:
Storing references as members is considered bad style and it can lead to side-effects when copying. `std::reference_wrapper` is essentially a `pointer` (internally) that gives us reference semantics. I don't like using raw pointers in general so `std::reference_wrapper` is probably the best option here imo.